### ISSUE TITLE
Failed tests appears as successful in TeamCity

### ISSUE THREAD

**Issue by DJ-Glock** [travisjeffery/mocha-teamcity-reporter#61]
## The problem
I use mocha-teamcity-reporter with the following options:
`--reporter-options recordHookFailures=true --reporter-options ignoreHookWithName=RootHook `

In the raw log I can see that test fails and messages are fine:
```
[36me2e_1  |[0m ##teamcity[testStarted name='FtsUI - 1.2 Security selected #full #intfull' captureStandardOutput='true' flowId='19']
[36me2e_1  |[0m ##teamcity[testStarted name='"before each" hook: Clean search in "FtsUI - 1 Search Input "' captureStandardOutput='true' flowId='19']
[36me2e_1  |[0m ##teamcity[testFinished name='"before each" hook: Clean search in "FtsUI - 1 Search Input "' flowId='19']
[36me2e_1  |[0m ##teamcity[testFailed name='FtsUI - 1.2 Security selected #full #intfull' message='Test and Master SecuritySelected.png files should be equal: expected |'Screenshot /e2e/test-screenshots/FTS/SecuritySelected.png did not match with master screenshot /e2e/master-screenshots/FTS/SecuritySelected. DiffBounds: {"left":0,"top":0,"right":180,"bottom":31} Diff-image will be saved to /e2e/diff-screenshots/FTS/SecuritySelected.png.|' to equal true' details='AssertionError: Test and Master SecuritySelected.png files should be equal: expected |'Screenshot /e2e/test-screenshots/FTS/SecuritySelected.png did not match with master screenshot /e2e/master-screenshots/FTS/SecuritySelected. DiffBounds: {"left":0,"top":0,"right":180,"bottom":31} Diff-image will be saved to /e2e/diff-screenshots/FTS/SecuritySelected.png.|' to equal true|n    at Function.compareScreenshots (src/Tests/UIGeneric.ts:21:80)|n    at processTicksAndRejections (internal/process/task_queues.js:93:5)|n    at Context.<anonymous> (src/TestsPoMa/FTS/FTS.prl.ui.ts:45:7)' captureStandardOutput='true' flowId='19']
[36me2e_1  |[0m ##teamcity[testFinished name='FtsUI - 1.2 Security selected #full #intfull' duration='6166' flowId='19']
```

But in TeamCity:
![image](https://user-images.githubusercontent.com/26511191/119131313-f4f0e280-ba41-11eb-9f3b-837ca1c2b7d4.png)

I suspect that TeamCity parses the following line and thinks that test finished:
`[36me2e_1  |[0m ##teamcity[testFinished name='"before each" hook: Clean search in "FtsUI - 1 Search Input "' flowId='19']`

This was changed in scope of https://github.com/travisjeffery/mocha-teamcity-reporter/issues/54 before that testFinished was not logged for hooks:
https://github.com/travisjeffery/mocha-teamcity-reporter/pull/58/files

I believe that the root cause of this issue is a but on TeamCity side - it looks like they find the next testFinished line and think that test is finished. But this line is for hook!

I will raise now an issue for TeamCity to check if it was fixed or not (we are using 2020.2.1 TC version and are not going to update it soon). So If they confirm that issue is on their side - I will create a PR to revert `hook end` implementation.

## Environment

* What mocha-teamcity-reporter version are you using: 4.0.0
* Version of teamcity (If Applicable): 2020.2.1 (build 85633)

**Comment by DJ-Glock** [comment]
I raised an issue for TeamCity - https://youtrack.jetbrains.com/issue/TW-71626

I run my build with the previous version of reporter and it works fine. So it looks like an issue on TC side.
I am going to create a PR to partially rollback changes made in https://github.com/travisjeffery/mocha-teamcity-reporter/pull/58/commits if they confirm the issue.

**Comment by DJ-Glock** [comment]
And also for some reason when I run tests **without** recordHookFailures flag - I get hook end testFinished messages.
Don't know why - when I run reporter tests they look fine. Will investigate it also.

**Comment by DJ-Glock** [comment]
OKay. TeamCity support was very helpful. They provided me a link to their doc. The issue here is in the same flowId for test and hook.

[Official doc](https://www.jetbrains.com/help/teamcity/service-messages.html#Nested+test+reporting) says:
_Starting another test finishes the currently started test in the same flow. To still report tests from within other tests, you will need to specify another flowId in the nested test service messages._

Ksenia from TC team advised using some prefix for example:
_You should use different flowids to distinguish separate processes running in parallel. You can use some prefix in the flowid for hooks, it should fix the problem_

So to solve this I believe we need to add some prefix to flowId for hooks. I'll check tomorrow and will do a PR.

**Comment by DJ-Glock** [comment]
- [ ] Fix an issue with treating failed test as successful
- [ ] Add tests to cover hooks reporting without recordHookFailures
- [ ] Check why recordHookFailures is true by default for my build

**Comment by DJ-Glock** [comment]
By the way I've just realized that `recordHookFailures` name does not really describe it's functionality. `fail` event is fired for failed hooks and tests hence it always logs failed hooks. :)

**Comment by DJ-Glock** [comment]
@jamie-sherriff please review https://github.com/travisjeffery/mocha-teamcity-reporter/pull/63 once able. I believe this fix can be very critical for users who use recordHookFailures option.

**Comment by jamie-sherriff** [comment]
@DJ-Glock Thanks for this, had a quick read of the teamcity docs, do you think its worthwhile to add parent=siuite1 etc on tests? Happy to break it into another issue if you think its worthwhile todo.

**Comment by DJ-Glock** [comment]
> @DJ-Glock Thanks for this, had a quick read of the teamcity docs, do you think its worthwhile to add parent=siuite1 etc on tests? Happy to break it into another issue if you think its worthwhile todo.

Mmm. I guess I didn't get it. What exactly do you want to cover?
Anyway I believe adding more tests is a good idea. It could be done in separate issue.

**Comment by DJ-Glock** [comment]
I see. https://www.jetbrains.com/help/teamcity/service-messages.html#Message+FlowId
Well I believe we should add this carefully in a separate issue.

**Comment by jamie-sherriff** [comment]
@DJ-Glock Yup fair enough, just bouncing some ideas. I don't use TeamCity on a daily basis anymore so am a bit out of the loop. Are you ok with this going out as a patch release or prefer a minor version bump?

**Comment by DJ-Glock** [comment]
@jamie-sherriff Minor version bump will be okay I believe.

**Comment by jamie-sherriff** [comment]
Closing as fixed by published version 4.1.0
